### PR TITLE
Simpler batch implementation.

### PR DESCRIPTION
**Pull request by jortel**
The `Batch` greatly simplified.

Fixed Download._repair logging stack trace for DownloadError (like 404s)

### CODE REVIEW

File: plugin/pulpcore/plugin/download/futures/batch.py [hunk 1]
```diff
@@ -52,33 +46,46 @@ class Batch:
...
+    def completed(self, plan):
         """
-        return self._is_shutdown
+        A plan has completed.
+
+        Args:
+            plan (Plan): The completed plan.
+        """
+        self.outstanding -= 1
+        for download in self.downloads:
```
Comment (mhrivnak) [inline plugin/pulpcore/plugin/download/futures/batch.py]: I don't think accessing an iterator is necessarily thread-safe, so it probably needs to acquire a lock here.

Comment (jortel) [reply plugin/pulpcore/plugin/download/futures/batch.py]: True.  good catch.

File: plugin/pulpcore/plugin/download/futures/batch.py [hunk 2]
```diff
@@ -89,29 +96,13 @@ def download(self):
                 The iterator will yield the download `Plan` in the order completed.
         """
         log.debug(_('%(batch)s - download started'), {'batch': self})
-        self.feeder.start()
+        for download in self.downloads:
+            plan = Plan(download)
+            self.submit(plan)
+            if self.outstanding == (self.concurrent + 1):
```
[the remark this answers is not in the document]
Comment (jortel) [reply plugin/pulpcore/plugin/download/futures/batch.py]: I like the proposed changes here and in your PR re: moving outstanding to the iterator.

Although, I don't see the value in re-introducing BACKLOG.  What's the value of having more items in the pipeline than we have threads in the executor?  If you're concerned about momentary starvation, we could limit the pipeline to `2 x concurrent`.  It's unlikely but what if BACKLOG = 1024 and concurrent = 2048?  I know it's not very realistic but demonstrates the added complexity of having 2 values.

[the remark this answers is not in the document]
Comment (mhrivnak) [reply plugin/pulpcore/plugin/download/futures/batch.py]: Good point. The starvation concern is limited to the fact that calling ``next`` on the iterator might itself be expensive. Consider:

```
def xml_iterator():
  for item in somehow_parse_an_xml_file():
    artifacts = do_expensive_processing(item)
    for artifact in artifacts:
      yield artifact.as_http_download()
```

If that iterator is being used to feed the downloads, it's beneficial to do some processing in advance and maintain a buffer of available downloads that are ready to use.

I like the idea of making the buffer size a multiple of the concurrency number. For a very small amount of memory, we could make it 4x, 5x, 10x the concurrency level. It's hard to prove what will work best or how much it matters, but my gut says we should have at least a few downloads ready to go for each thread at any given time.

File: plugin/pulpcore/plugin/download/futures/batch.py [hunk 3]
```diff
@@ -89,29 +96,13 @@ def download(self):
                 The iterator will yield the download `Plan` in the order completed.
         """
         log.debug(_('%(batch)s - download started'), {'batch': self})
-        self.feeder.start()
+        for i, download in enumerate(self.downloads):
```
Comment (mhrivnak) [inline plugin/pulpcore/plugin/download/futures/batch.py]: If getting downloads out of this iterator is even slightly expensive (as described in another comment), it's possible that the ``completed()`` method above might get called before this iteration is finished. So I think the lock should be acquired here as well.

Comment (jortel) [reply plugin/pulpcore/plugin/download/futures/batch.py]: Agreed.  Done.

**Pull request merged**